From 9f0469ecf82e5248104a2564abf2dab896311768 Mon Sep 17 00:00:00 2001
From: Antonio Borneo <antonio.borneo@foss.st.com>
Date: Wed, 8 Feb 2023 11:01:10 +0100
Subject: [PATCH 0341/1695] drivers: stm32_exti: remove r/w data from struct
 stm32_exti_bank

To prepare the driver for multi instances, split away from struct
stm32_exti_bank the R/W fields used for instance data and move
them in the private data struct stm32_exti_pdata.
Make const all the instances of struct stm32_exti_bank.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 801a3b01ac1be52d7ac75d8ade1e5d23d9971bcc drivers: stm32_exti: remove r/w data from struct stm32_exti_bank]
Signed-off-by: Antonio Borneo <antonio.borneo@foss.st.com>
Change-Id: If14226940c8d7ed4fdce36665d2be21ab9132b05
---
 core/drivers/stm32_exti.c | 125 +++++++++++++++-----------------------
 1 file changed, 49 insertions(+), 76 deletions(-)

diff --git a/core/drivers/stm32_exti.c b/core/drivers/stm32_exti.c
index 3a5727cf7..fc02b6cd2 100644
--- a/core/drivers/stm32_exti.c
+++ b/core/drivers/stm32_exti.c
@@ -26,21 +26,9 @@ struct stm32_exti_bank {
 	uint32_t rpr_ofst;
 	uint32_t fpr_ofst;
 	uint32_t tzenr_ofst;
-	uint32_t wakeup;
-	uint32_t imr_cache;
-	uint32_t rtsr_bkp;
-	uint32_t ftsr_bkp;
-	uint32_t tz_bkp;
 };
 
-struct stm32_exti_pdata {
-	vaddr_t base;
-	uint32_t port_sel_cache[_EXTI_MAX_CR];
-};
-
-static struct stm32_exti_pdata stm32_exti;
-
-static struct stm32_exti_bank stm32mp1_exti_b1 = {
+static const struct stm32_exti_bank stm32mp1_exti_b1 = {
 	.imr_ofst	= 0x80,
 	.rtsr_ofst	= 0x00,
 	.ftsr_ofst	= 0x04,
@@ -49,7 +37,7 @@ static struct stm32_exti_bank stm32mp1_exti_b1 = {
 	.tzenr_ofst	= 0x14,
 };
 
-static struct stm32_exti_bank stm32mp1_exti_b2 = {
+static const struct stm32_exti_bank stm32mp1_exti_b2 = {
 	.imr_ofst	= 0x90,
 	.rtsr_ofst	= 0x20,
 	.ftsr_ofst	= 0x24,
@@ -58,7 +46,7 @@ static struct stm32_exti_bank stm32mp1_exti_b2 = {
 	.tzenr_ofst	= 0x34,
 };
 
-static struct stm32_exti_bank stm32mp1_exti_b3 = {
+static const struct stm32_exti_bank stm32mp1_exti_b3 = {
 	.imr_ofst	= 0xA0,
 	.rtsr_ofst	= 0x40,
 	.ftsr_ofst	= 0x44,
@@ -67,12 +55,26 @@ static struct stm32_exti_bank stm32mp1_exti_b3 = {
 	.tzenr_ofst	= 0x54,
 };
 
-static struct stm32_exti_bank *stm32mp1_exti_banks[] = {
+static const struct stm32_exti_bank *stm32mp1_exti_banks[] = {
 	&stm32mp1_exti_b1,
 	&stm32mp1_exti_b2,
 	&stm32mp1_exti_b3,
 };
 
+#define _EXTI_BANK_NR ARRAY_SIZE(stm32mp1_exti_banks)
+
+struct stm32_exti_pdata {
+	vaddr_t base;
+	uint32_t wake_active[_EXTI_BANK_NR];
+	uint32_t mask_cache[_EXTI_BANK_NR];
+	uint32_t rtsr_cache[_EXTI_BANK_NR];
+	uint32_t ftsr_cache[_EXTI_BANK_NR];
+	uint32_t seccfgr_cache[_EXTI_BANK_NR];
+	uint32_t port_sel_cache[_EXTI_MAX_CR];
+};
+
+static struct stm32_exti_pdata stm32_exti;
+
 /* EXTI access protection */
 static unsigned int lock = SPINLOCK_UNLOCK;
 
@@ -143,7 +145,7 @@ void stm32_exti_mask(uint32_t exti_line)
 	val = io_read32(stm32_exti.base + stm32mp1_exti_banks[i]->imr_ofst);
 	val &= ~mask;
 	io_write32(stm32_exti.base + stm32mp1_exti_banks[i]->imr_ofst, val);
-	stm32mp1_exti_banks[i]->imr_cache = val;
+	stm32_exti.mask_cache[i] = val;
 
 	cpu_spin_unlock_xrestore(&lock, exceptions);
 }
@@ -162,7 +164,7 @@ void stm32_exti_unmask(uint32_t exti_line)
 	val = io_read32(stm32_exti.base + stm32mp1_exti_banks[i]->imr_ofst);
 	val |= mask;
 	io_write32(stm32_exti.base + stm32mp1_exti_banks[i]->imr_ofst, val);
-	stm32mp1_exti_banks[i]->imr_cache = val;
+	stm32_exti.mask_cache[i] = val;
 
 	cpu_spin_unlock_xrestore(&lock, exceptions);
 }
@@ -177,7 +179,7 @@ void stm32_exti_enable_wake(uint32_t exti_line)
 
 	exceptions = cpu_spin_lock_xsave(&lock);
 
-	stm32mp1_exti_banks[i]->wakeup |= mask;
+	stm32_exti.wake_active[i] |= mask;
 
 	cpu_spin_unlock_xrestore(&lock, exceptions);
 }
@@ -192,7 +194,7 @@ void stm32_exti_disable_wake(uint32_t exti_line)
 
 	exceptions = cpu_spin_lock_xsave(&lock);
 
-	stm32mp1_exti_banks[i]->wakeup &= ~mask;
+	stm32_exti.wake_active[i] &= ~mask;
 
 	cpu_spin_unlock_xrestore(&lock, exceptions);
 }
@@ -251,36 +253,21 @@ static void stm32_exti_pm_suspend(void)
 	uint32_t base = stm32_exti.base;
 	uint32_t mask = 0;
 	uint32_t i = 0;
+	const struct stm32_exti_bank *bank = NULL;
+
+	for (i = 0; i < _EXTI_BANK_NR; i++) {
+		bank = stm32mp1_exti_banks[i];
 
-	/* Save ftsr, rtsr and tzen registers */
-	stm32mp1_exti_b1.ftsr_bkp = io_read32(stm32_exti.base +
-					      stm32mp1_exti_b1.ftsr_ofst);
-	stm32mp1_exti_b1.rtsr_bkp = io_read32(stm32_exti.base +
-					      stm32mp1_exti_b1.rtsr_ofst);
-	stm32mp1_exti_b1.tz_bkp = io_read32(stm32_exti.base +
-					    stm32mp1_exti_b1.tzenr_ofst);
-
-	stm32mp1_exti_b2.ftsr_bkp = io_read32(stm32_exti.base +
-					      stm32mp1_exti_b2.ftsr_ofst);
-	stm32mp1_exti_b2.rtsr_bkp = io_read32(stm32_exti.base +
-					      stm32mp1_exti_b2.rtsr_ofst);
-	stm32mp1_exti_b2.tz_bkp = io_read32(stm32_exti.base +
-					    stm32mp1_exti_b2.tzenr_ofst);
-
-	stm32mp1_exti_b3.ftsr_bkp = io_read32(stm32_exti.base +
-					      stm32mp1_exti_b3.ftsr_ofst);
-	stm32mp1_exti_b3.rtsr_bkp = io_read32(stm32_exti.base +
-					      stm32mp1_exti_b3.rtsr_ofst);
-	stm32mp1_exti_b3.tz_bkp = io_read32(stm32_exti.base +
-					    stm32mp1_exti_b3.tzenr_ofst);
-
-	/* Let enabled only the wakeup sources set*/
-	mask = stm32mp1_exti_b1.wakeup;
-	io_mask32(stm32_exti.base + stm32mp1_exti_b1.imr_ofst, mask, mask);
-	mask = stm32mp1_exti_b2.wakeup;
-	io_mask32(stm32_exti.base + stm32mp1_exti_b2.imr_ofst, mask, mask);
-	mask = stm32mp1_exti_b3.wakeup;
-	io_mask32(stm32_exti.base + stm32mp1_exti_b3.imr_ofst, mask, mask);
+		/* Save ftsr, rtsr and seccfgr registers */
+		stm32_exti.ftsr_cache[i] = io_read32(base + bank->ftsr_ofst);
+		stm32_exti.rtsr_cache[i] = io_read32(base + bank->rtsr_ofst);
+		stm32_exti.seccfgr_cache[i] =
+			io_read32(base + bank->tzenr_ofst);
+
+		/* Let enabled only the wakeup sources set */
+		mask = stm32_exti.wake_active[i];
+		io_mask32(base + bank->imr_ofst, mask, mask);
+	}
 
 	/* Save EXTI port selection */
 	for (i = 0; i < _EXTI_MAX_CR; i++)
@@ -291,34 +278,20 @@ static void stm32_exti_pm_resume(void)
 {
 	uint32_t base = stm32_exti.base;
 	uint32_t i = 0;
+	const struct stm32_exti_bank *bank = NULL;
+
+	for (i = 0; i < _EXTI_BANK_NR; i++) {
+		bank = stm32mp1_exti_banks[i];
 
-	/* Restore ftsr and rtsr registers */
-	io_write32(stm32_exti.base + stm32mp1_exti_b1.ftsr_ofst,
-		   stm32mp1_exti_b1.ftsr_bkp);
-	io_write32(stm32_exti.base + stm32mp1_exti_b2.ftsr_ofst,
-		   stm32mp1_exti_b2.ftsr_bkp);
-	io_write32(stm32_exti.base + stm32mp1_exti_b3.ftsr_ofst,
-		   stm32mp1_exti_b3.ftsr_bkp);
-	io_write32(stm32_exti.base + stm32mp1_exti_b1.rtsr_ofst,
-		   stm32mp1_exti_b1.rtsr_bkp);
-	io_write32(stm32_exti.base + stm32mp1_exti_b2.rtsr_ofst,
-		   stm32mp1_exti_b2.rtsr_bkp);
-	io_write32(stm32_exti.base + stm32mp1_exti_b3.rtsr_ofst,
-		   stm32mp1_exti_b3.rtsr_bkp);
-	io_write32(stm32_exti.base + stm32mp1_exti_b1.tzenr_ofst,
-		   stm32mp1_exti_b1.tz_bkp);
-	io_write32(stm32_exti.base + stm32mp1_exti_b2.tzenr_ofst,
-		   stm32mp1_exti_b2.tz_bkp);
-	io_write32(stm32_exti.base + stm32mp1_exti_b3.tzenr_ofst,
-		   stm32mp1_exti_b3.tz_bkp);
-
-	/* Restore imr */
-	io_write32(stm32_exti.base + stm32mp1_exti_b1.imr_ofst,
-		   stm32mp1_exti_b1.imr_cache);
-	io_write32(stm32_exti.base + stm32mp1_exti_b2.imr_ofst,
-		   stm32mp1_exti_b2.imr_cache);
-	io_write32(stm32_exti.base + stm32mp1_exti_b3.imr_ofst,
-		   stm32mp1_exti_b3.imr_cache);
+		/* Restore ftsr, rtsr and seccfgr registers */
+		io_write32(base + bank->ftsr_ofst, stm32_exti.ftsr_cache[i]);
+		io_write32(base + bank->rtsr_ofst, stm32_exti.rtsr_cache[i]);
+		io_write32(base + bank->tzenr_ofst,
+			   stm32_exti.seccfgr_cache[i]);
+
+		/* Restore imr */
+		io_write32(base + bank->imr_ofst, stm32_exti.mask_cache[i]);
+	}
 
 	/* Restore EXTI port selection */
 	for (i = 0; i < _EXTI_MAX_CR; i++)
-- 
2.47.2

